Add program_name connectionAttribute for MySQL

jeff303 opened the pull request:
Set the `program_name` `connectionAttribute` when creating a MySQL connection, to the same value that we do for other drivers - `config/mb-version-and-process-identifier`

Updating tests

Relevant links:

* https://dev.mysql.com/doc/refman/8.0/en/performance-schema-connection-attribute-tables.html#performance-schema-connection-attributes-available
* https://mariadb.com/kb/en/performance-schema-session_connect_attrs-table/
* http://mysqlblog.fivefarmers.com/2013/03/26/connection-attributes-in-mysql-5-6/

jeff303 committed:
Add program_name connectionAttribute for MySQL

Set the program_name connection attribute when creating a MySQL connection, to the same value that we do for other drivers - config/mb-version-and-process-identifier

Updating tests

Relevant links:

* https://dev.mysql.com/doc/refman/8.0/en/performance-schema-connection-attribute-tables.html#performance-schema-connection-attributes-available
* https://mariadb.com/kb/en/performance-schema-session_connect_attrs-table/
* http://mysqlblog.fivefarmers.com/2013/03/26/connection-attributes-in-mysql-5-6/

pawit-metabase commented on src/metabase/driver/mysql.clj:
```diff
@@ -332,7 +342,7 @@
                        (dissoc :ssl))]
        (-> (dbspec/mysql details)
-           (sql-jdbc.common/handle-additional-options details))))))
+           (sql-jdbc.common/handle-additional-options (add-program-name-option details)))))))
```
```suggestion
           (sql-jdbc.common/handle-additional-options details)
           (assoc :connectionAttributes (str "program_name:" (str/replace config/mb-version-and-process-identifier "," "_"))))))))
```

Rather than messing with "additional options", you can just add it to the spec which will end up in JDBC properties and the driver will pick it up.

The above suggestion is just a simple example of how to make it work without changing additional options. Probably you'd still want the logic to be in `add-program-name-option`.

flamber replied:
I don't fully understand the code, but it would be very helpful if we allow additional options to override our hardcoded stuff. We've had several issues, where we hardcode something, which means it cannot be overridden.

jeff303 replied:
@pawit-metabase , done.  That is a little cleaner.

@flamber , this is just the string that basically shows up in the DB if you query for active connections.  See the 2nd comment above ^.  I am open to allowing override of this string, for very sophisticated users (most likely), but this particular PR is doing nothing more than extending the work already done for Oracle and Postgres, to MySQL (see #15679).  In any event, this should be a vast improvement over seeing either an empty string, or some generic name like `connector/j` or whatever.

jeff303 committed:
PR feedback

Set :connectionAttributes directly in the jdbc-spec map (which in turn becomes a Property for the JDBC connection) instead of putting in additional-options

camsaul commented on src/metabase/driver/mysql.clj:
```diff
@@ -313,6 +314,12 @@
+  ;; connectionAttributes (if multiple) are separated by commas, so values that contain spaces are OK, so long as they
+  ;; don't contain a comma; our mb-version-and-process-identifier shouldn't contain one, but just to be on the safe side
+  (let [prog-name (str/replace config/mb-version-and-process-identifier "," "_")]
+    (assoc jdbc-spec :connectionAttributes (str "program_name:" prog-name))))
```
If someone specifies `connectionAttributes` in the additional JDBC options, this would override that, wouldn't it? Are there any other `connectionAttributes` one might specify? I wouldn't want to break things if someone had been setting this in the past and then we started stomping on that

camsaul replied:
You might want to look for that and not set it if it's in the additional options

jeff303 replied:
This is now handled (new test added)

camsaul reviewed (approved):
I'm 👍 on this as long as we're considering what would happen if someone specified this in the additional connection options and not stomping on that

jeff303 committed:
Merge branch 'master' into add-program-name-mysql-connection-string

jeff303 committed:
Merge branch 'master' into add-program-name-mysql-connection-string

jeff303 committed:
PR feedback

Fix some bugs in additional-options->map

Add test for mysql to ensure :connectionAttributes aren't emitted when they are given in additional-options

jeff303 committed:
Merge branch 'master' into add-program-name-mysql-connection-string
